### DISCUSSION TITLE
prepare and prepublish not recognized

### DISCUSSION THREAD

**Discussion by steven-coding** [naugtur/can-i-ignore-scripts#13] [General]
Why are those lifecycle hooks not recognized?

In the source code I have seen:
`const scriptNames = ['preinstall', 'postinstall', 'install']`
So those scripts are recognized.

But `prepublish` and `prepare` are called during `npm install` and could also trigger harmful scripts to be run.
They don't seem to be recognized by this package.

Or am I wrong here?

**Comment by naugtur** [comment]
I'm gonna have to set up a whole test suite to check for sure and detect if any versions did otherwise, but as far as I know, npm ci or nom install is only running those for local packages you link install from folders or other weird sources and _not_ when i stalling from the registry.

Did you base your question on the docs or actually seeing the scripts?

**Reply by steven-coding** [reply]
Yesterday I found the following in the docs:
https://docs.npmjs.com/cli/v7/using-npm/scripts#npm-install

> These also run when you run npm install -g \<pkg-name\>
> 
> preinstall
> install
> postinstall
> prepublish
> preprepare
> prepare
> postprepare


I checked it on a new (local) package with the following package.json:
```
{
  "name": "npm-test",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
	"preinstall": "echo preinstall called!",
	"install": "echo install called!",
	"postinstall": "echo postinstall called!",
	"prepublish": "echo prepublish called!",
	"preprepare": "echo preprepare called!",
	"prepare": "echo prepare called!",
	"postprepare": "echo postprepare called!"
  },
  "author": "",
  "license": "ISC"
}
```

and did `npm install` within this package.
It created the following output:
```
PS C:\Development\npm-test> npm install

> npm-test@1.0.0 preinstall C:\Development\npm-test
> echo preinstall called!

preinstall called!

> npm-test@1.0.0 install C:\Development\npm-test
> echo install called!

install called!

> npm-test@1.0.0 postinstall C:\Development\npm-test
> echo postinstall called!

postinstall called!
npm WARN prepublish-on-install As of npm@5, `prepublish` scripts are deprecated.
npm WARN prepublish-on-install Use `prepare` for build steps and `prepublishOnly` for upload-only.
npm WARN prepublish-on-install See the deprecation note in `npm help scripts` for more information.

> npm-test@1.0.0 prepublish C:\Development\npm-test
> echo prepublish called!

prepublish called!

> npm-test@1.0.0 prepare C:\Development\npm-test
> echo prepare called!

prepare called!
npm notice created a lockfile as package-lock.json. You should commit this file.
npm WARN npm-test@1.0.0 No description
npm WARN npm-test@1.0.0 No repository field.

up to date in 0.627s
found 0 vulnerabilities
```

Ergo the following 5 lifecycle-scripts were triggered on `npm install`:
- preinstall
- install
- postinstall
- prepublish
- prepare

I assume that this will also be the case for dependency packages, but I have to check this.

**Reply by steven-coding** [reply]
I was able to cross-check it as a dependency install via github and my assumption was wrong.
It triggered:
- preinstall
- install
- postinstall

of the dependency when calling `npm install` on the consuming project.

So your cases are completely valid and intact.
I simply misunderstood the documentation of npm.

Sorry for any inconvenience!

**Reply by naugtur** [reply]
No problem, I went through the same process, just a bunch of months earlier ;)
